### ISSUE TITLE
S1172 Unused parameters for method syntax defined by delegate

### ISSUE THREAD

**Issue by pedromagueija** [SonarSource-VisualStudio/sonarlint-roslyn-analyzer#632]
There is a false positive reported by SonarLint when an parameter is not used in a method that is used as a delegate e.g.: an event handler.

```C#
namespace ConsoleApplication2
{
    using System;

    internal class Program
    {
        private static void Main(string[] args)
        {
            SomeClass instance= new SomeClass();
            instance.SomeEvent += OnSomeEvent;    
        }

        // arg1 is marked as unused (true, but necessary)
        private static void OnSomeEvent(string arg1, int arg2)
        {
            if (arg2 == 0)
            {
                Console.WriteLine("its zero");
            }
        }
    }

    class SomeClass
    {
        public event CustomDelegate SomeEvent;
    }

    delegate void CustomDelegate(string arg1, int arg2);
}
```

**Comment by tvsonar** [comment]
@pedromagueija thanks for reporting this issue. I think in this case `arg1` is unused, and it's not necessary, because you could change your custom delegate to have only one parameter.

We have a mechanism to recognize event handlers, but that only works for delegates that conform to the [event handler definition guidelines](https://msdn.microsoft.com/library/ms229011(v=vs.100).aspx). (Two parameters, the first is `object`, the second one `EventArgs`). This was handled in #211.

Do you think we should exclude other patterns as well? What's the reason that you are not using the standard [`EventHandler<T>`](https://msdn.microsoft.com/en-us/library/db0etb8x(v=vs.110).aspx) delegate?

**Comment by pedromagueija** [comment]
Hi @tvsonar,

I missed to mentioned that the event (and the delegate) in question are from a 3rd party API. I have no control over the source (proprietary API).

The example I posted is what I assume internally is done in that API.

Sorry about that.

Cheers.

**Comment by tvsonar** [comment]
For private/internal methods we could do a bit more: we can check if the method in question is being assigned to an event anywhere in the class, and if so, then we could exclude it.

**Comment by pedromagueija** [comment]
Is it possible to check if the method is being assigned to an event or delegate defined outside the assembly, and in that case exclude it from this check?

**Comment by tvsonar** [comment]
Yes, we can check if the event is declared outside the assembly.
